phpBB/assets/javascript/editor.js

```diff
@@ -24,337 +6,382 @@ function helpline(help) {
+	editor.insertBBCode(e.target.alt);
+	e.preventDefault();
+});
 
-	var textarea = document.forms[form_name].elements[text_name];
 
-	textarea.focus();
+var editorConstants = {
+	'NO_VALUE' 			: -1,
+	'VALUE_IN_CONTENT' 	: -2,
 
-	if ((clientVer >= 4) && is_ie && is_win) {
-		// Get text selection
-		theSelection = document.selection.createRange().text;
+};
 
-		if (theSelection) {
-			// Add tags around selection
-			document.selection.createRange().text = bbopen + theSelection + bbclose;
-			textarea.focus();
-			theSelection = '';
-			return;
-		}
-	} else if (textarea.selectionEnd && (textarea.selectionEnd - textarea.selectionStart > 0)) {
-		mozWrap(textarea, bbopen, bbclose);
-		textarea.focus();
-		theSelection = '';
-		return;
-	}
+var editor = {
 
-	//The new position for the cursor after adding the bbcode
-	var caret_pos = getCaretPosition(textarea).start;
-	var new_pos = caret_pos + bbopen.length;
+	tokenRegex: {
+		'ALPHANUM': /^[0-9A-Za-z]+$/,
+		'SIMPLETEXT': /^[a-z0-9,.\-+_]+$/i,
+		'IDENTIFIER': /^[a-z0-9-_]+$/i,
+		'INTTEXT': /^[a-zA-Z\u00C0-\u017F]+,\s[a-zA-Z\u00C0-\u017F]+$/,
+		'NUMBER': /^[0-9]+$/,
+		'INTEGER': /^(?:0|-?[1-9]\d*)$/,
 
-	// Open tag
-	insert_text(bbopen + bbclose);
+		'EMAIL': /[a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+\/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+(?:[A-Z]{2}|com|org|net|edu|gov|mil|me|biz|info|mobi|name|aero|asia|jobs|museum)\b/,
 
-	// Center the cursor when we don't have a selection
-	// Gecko and proper browsers
-	if (!isNaN(textarea.selectionStart)) {
-		textarea.selectionStart = new_pos;
-		textarea.selectionEnd = new_pos;
-	}
-	// IE
-	else if (document.selection) {
-		var range = textarea.createTextRange(); 
-		range.move("character", new_pos); 
-		range.select();
-		storeCaret(textarea);
-	}
+		'URL': /^(?:(?:https?|ftps?):\/\/)?(?:(?:[a-z]+@)?(([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})|((?:[A-F0-9]{1,4}:){7}[A-F0-9]{1,4})|((?:[A-F0-9]{1,4}:){1,4}:(?:[A-F0-9]{1,4}:){0,4}[A-F0-9]{1,4})|(?:(?:(?:[a-z0-9-]|%\d\d)+\.)+[a-z]{2,7})|localhost))?(?:\/([a-z0-9-\/.]*))?(?:\?((?:[^=]+=[^&]+&)*(?:[^=]+=[^#]+)?))?(?:#.*)?$/im,
+		'LOCAL_URL': /^(?:\/([a-z0-9-\/.]*))?(?:\?((?:[^=]+=[^&]+&)*(?:[^=]+=[^#$]+)?))?(?:#[^$]*)?$/,
+		'RELATIVE_URL': /^(?:\/([a-z0-9-\/.]*))?(?:\?((?:[^=]+=[^&]+&)*(?:[^=]+=[^#$]+)?))?(?:#[^$]*)?$/,
 
-	textarea.focus();
-	return;
-}
+		'COLOR': /^(?:#[0-9a-f]{3,6}|rgb\(\d{1,3}, *\d{1,3}, *\d{1,3}\)|aqua|black|blue|fuchsia|gray|green|lime|maroon|navy|olive|orange|purple|red|silver|teal|white|yellow)$/i
+	},
 
-/**
-* Insert text at position
-*/
-function insert_text(text, spaces, popup) {
-	var textarea;
+	getElementDefaultDisplay: function () {
+		if(window.getComputedStyle){
+			return function (tagName){
+				var tag = document.createElement(tagName);
+				document.body.appendChild(tag);
+				var cStyle = window.getComputedStyle(tag, "").display;
+				document.body.removeChild(tag);
+				return cStyle;
```

Comment:
jQuery does extra work to get values and, sometimes, doesn't get the value I want and gets something else instead. I want it to fail if it would be getting the value through those alternative ways.